delta-sharing/core/src/capabilities.rs

```diff
@@ -0,0 +1,179 @@
+    /// Create a new [`Capabilities`] instance.
+    ///
+    /// # Example
+    /// ```
+    /// use delta_sharing_core::capabilities::{Capabilities, ResponseFormat};
+    ///
+    /// let capabilities = Capabilities::new(
+    ///   vec![ResponseFormat::Delta],
+    ///   vec!["deletionVectors".to_string()],
+    /// );
+    /// assert_eq!(capabilities.response_formats(), &[ResponseFormat::Delta]);
+    /// ```
+    pub fn new(response_formats: Vec<ResponseFormat>, reader_features: Vec<String>) -> Self {
+        Self {
+            response_formats,
+            reader_features: reader_features
+                .into_iter()
+                .map(|s| s.to_lowercase())
+                .collect(),
+        }
+    }
+
+    /// Returns the response formats that the client can accept.
+    ///
+    /// # Example
+    /// ```
+    /// use delta_sharing_core::capabilities::{Capabilities, ResponseFormat};
+    ///
+    /// let capabilities = Capabilities::new(
+    ///   vec![ResponseFormat::Delta],
+    ///   vec!["deletionVectors".to_string()],
+    /// );
+    /// assert_eq!(capabilities.response_formats(), &[ResponseFormat::Delta]);
+    /// ```
+    pub fn response_formats(&self) -> &[ResponseFormat] {
+        &self.response_formats
+    }
+
+    /// Returns the reader features that the client can accept.
+    ///
+    /// # Example
+    /// ```
+    /// use delta_sharing_core::capabilities::{Capabilities, ResponseFormat};
+    ///
+    /// let capabilities = Capabilities::new(
+    ///   vec![ResponseFormat::Delta],
+    ///   vec!["deletionVectors".to_string()],
+    /// );
+    /// assert_eq!(capabilities.reader_features(), &["deletionvectors"]);
+    pub fn reader_features(&self) -> &[String] {
+        self.reader_features.as_slice()
+    }
+}
+
+impl Default for Capabilities {
+    fn default() -> Self {
+        Self {
+            response_formats: vec![ResponseFormat::Parquet],
+            reader_features: vec![],
+        }
+    }
+}
+
+impl TryFrom<&HeaderMap> for Capabilities {
+    type Error = Error;
+
+    fn try_from(headers: &HeaderMap) -> Result<Self, Self::Error> {
+        let mut capabilities = Capabilities::default();
+        if let Some(header) = headers.get(DELTA_SHARING_CAPABILITIES) {
+            let capability_header = header.to_str().map_err(|e| {
+                Error::Generic(format!("Failed to parse capabilities header: {}", e))
+            })?;
+            for capability in capability_header.split(';') {
+                let (capability_key, capability_value) = capability.split_once('=').ok_or(
+                    Error::Generic(format!("Failed to parse capability: {}", capability)),
+                )?;
+                match capability_key {
+                    "responseformat" => {
+                        capabilities.response_formats = capability_value
+                            .split(',')
+                            .flat_map(|s| s.trim().to_lowercase().parse::<ResponseFormat>().ok())
```

Comment:
nit: would it make sense to move the lower casing into the parse function such that all future code paths we may have handle that as well?

Reply:
good spot! Should be resolved now:)